crates/enarx-config/src/lib.rs

```diff
@@ -206,65 +213,89 @@ pub enum File {
-    #[serde(rename = "connect")]
-    Connect {
+/// File descriptor of a stream socket
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(tag = "prot", deny_unknown_fields)]
+pub enum ConnectFile {
+    /// TLS stream socket
+    #[serde(rename = "tls")]
+    Tls {
         /// Name assigned to the file descriptor
         name: Option<FileName>,
 
         /// Host address to connect to
         host: String,
 
         /// Port to connect to
-        #[serde(default = "default_port")]
+        #[serde(default = "default_tls_port")]
         port: u16,
+    },
+
+    /// TCP stream socket
+    #[serde(rename = "tcp")]
+    Tcp {
+        /// Name assigned to the file descriptor
+        name: Option<FileName>,
 
-        /// Protocol to use
-        #[serde(default)]
-        prot: Protocol,
+        /// Host address to connect to
+        host: String,
+
+        /// Port to connect to
+        #[serde(default = "default_tcp_port")]
+        port: u16,
     },
 }
 
+/// Parameters for a pre-opened file descriptor
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(tag = "kind", deny_unknown_fields)]
+pub enum File {
+    /// File descriptor of `/dev/null`
+    #[serde(rename = "null")]
+    Null(NullFile),
+
+    /// File descriptor of stdin
+    #[serde(rename = "stdin")]
+    Stdin(StdioFile),
+
+    /// File descriptor of stdout
+    #[serde(rename = "stdout")]
+    Stdout(StdioFile),
+
+    /// File descriptor of stderr
+    #[serde(rename = "stderr")]
+    Stderr(StdioFile),
+
+    /// File descriptor of a listen socket
+    #[serde(rename = "listen")]
+    Listen(ListenFile),
+
+    /// File descriptor of a stream socket
+    #[serde(rename = "connect")]
+    Connect(ConnectFile),
+}
+
 impl File {
     /// Get the name for a file descriptor
     pub fn name(&self) -> &str {
         match self {
-            Self::Null { name } => name.as_deref().unwrap_or("null"),
-            Self::Stdin { name } => name.as_deref().unwrap_or("stdin"),
-            Self::Stdout { name } => name.as_deref().unwrap_or("stdout"),
-            Self::Stderr { name } => name.as_deref().unwrap_or("stderr"),
-            Self::Listen { name, .. } => name,
-            Self::Connect { name, host, .. } => name.as_deref().unwrap_or(host),
+            Self::Null(NullFile { name }) => name.as_deref().unwrap_or("null"),
+            Self::Stdin(StdioFile { name }) => name.as_deref().unwrap_or("stdin"),
+            Self::Stdout(StdioFile { name }) => name.as_deref().unwrap_or("stdout"),
+            Self::Stderr(StdioFile { name }) => name.as_deref().unwrap_or("stderr"),
+            Self::Listen(ListenFile::Tls { name, .. }) => name,
+            Self::Listen(ListenFile::Tcp { name, .. }) => name,
+            Self::Connect(ConnectFile::Tls { name, host, .. }) => name.as_deref().unwrap_or(host),
+            Self::Connect(ConnectFile::Tcp { name, host, .. }) => name.as_deref().unwrap_or(host),
```

Comment:
Shouldn't the name also contain the port? Otherwise we might end up with multiple duplicate names.

Reply:
This method will be removed in next iteration and that's exactly how it worked before, there are no changes in behavior here.

Reply:
Files will be a hashmap, where name will be required and duplicates won't be allowed